### DISCUSSION TITLE
How can I bind value in Adorner?

### DISCUSSION THREAD

**Discussion by kyuranger** [AvaloniaUI/Avalonia#10776] [Q&A] [answered]
I want to bind the IsVisible value of the Adorner to IsChecked property of a usercontrol that inherits ToggleButton.

Here is my XAML code:
```
<UserControl xmlns="https://github.com/avaloniaui"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
             mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
			 xmlns:local="using:AvaloniaApplication1"
             x:Class="AvaloniaApplication1.UserControl1">
	<UserControl.Styles>
		<Style Selector=":is(local|UserControl1)">
			<Setter Property="Template">
				<ControlTemplate>
					<Border x:Name="B">
						<ContentPresenter  Content="{TemplateBinding Content}"></ContentPresenter>
						<AdornerLayer.Adorner>
							<Grid IsVisible="{Binding $parent[local|UserControl1].IsChecked}">
							</Grid>
						</AdornerLayer.Adorner>
					</Border>
				</ControlTemplate>
			</Setter>
		</Style>
	</UserControl.Styles>
</UserControl>

```

Code-Behind:
```
using Avalonia;
using Avalonia.Controls;
using Avalonia.Controls.Primitives;
using Avalonia.Markup.Xaml;
using System;

namespace AvaloniaApplication1;

public partial class UserControl1 : UserControl
{
    public UserControl1()
    {
        InitializeComponent();
    }
    public static readonly StyledProperty<Boolean> IsCheckedProperty =
AvaloniaProperty.Register<UserControl, Boolean>(nameof(IsChecked));

    public Boolean IsChecked
    {
        get { return GetValue(IsCheckedProperty); }
        set
        {           
            SetValue(IsCheckedProperty, value);
        }
    }
}
```

After the program ran, it reports an error:

>   Severity	Code	Description	Project	File	Line	Suppression State
Error	AVLN:0004	Unable to resolve type local from namespace https://github.com/avaloniaui Line 15, position 14.	AvaloniaApplication1	C:\Users\Desktop\AvaloniaApplication1\UserControl1.axaml	15	

It seems can not find the local xmln namespace. What's wrong with it?

**Comment by timunie** [comment, marked as answer]
A UserControl that inherits ToggleButton? That is not possible to do. If you want to create your own custom control, please see this sample: https://github.com/AvaloniaUI/Avalonia.Samples/tree/main/src/Avalonia.Samples/CustomControls/RatingControlSample

**Reply by kyuranger** [reply]
> A UserControl that inherits ToggleButton? That is not possible to do. If you want to create your own custom control, please see this sample: https://github.com/AvaloniaUI/Avalonia.Samples/tree/main/src/Avalonia.Samples/CustomControls/RatingControlSample

I apologize for didn't recreate a sample project, but directly copied and modified the code. Usercontrol cannot be inherited from ToggleButton and actually I created the property IsChecked by dependencyproperty. Now I recreated a sample project and updated all the code of my question again.

Thanks for telling me the sample tutorial. However, what I want is to bind a dependencyproperty in Adorner while the tutorial tell nothing about Adorner at all.

**Reply by timunie** [reply]
$parent is looking in the LogicalTree. I don't know if the Adorner is a child of it. You can use DevTools via [F12] to check Logical tree. Another idea could be to give the UserControl a name and bind to it via x:Name

```xml
<UserControl xmlns="https://github.com/avaloniaui"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
             mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
			 xmlns:local="using:AvaloniaApplication1"
             x:Class="AvaloniaApplication1.UserControl1"
             x:Name="root" >
	<UserControl.Styles>
		<Style Selector=":is(local|UserControl1)">
			<Setter Property="Template">
				<ControlTemplate>
					<Border x:Name="B">
						<ContentPresenter  Content="{TemplateBinding Content}"></ContentPresenter>
						<AdornerLayer.Adorner>
							<Grid IsVisible="{Binding #root.IsChecked}">
							</Grid>
						</AdornerLayer.Adorner>
					</Border>
				</ControlTemplate>
			</Setter>
		</Style>
	</UserControl.Styles>
</UserControl>
```

**Reply by kyuranger** [reply]
> $parent is looking in the LogicalTree. I don't know if the Adorner is a child of it. You can use DevTools via [F12] to check Logical tree. Another idea could be to give the UserControl a name and bind to it via x:Name
> 
> ```
> <UserControl xmlns="https://github.com/avaloniaui"
>              xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
>              xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
>              xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
>              mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
> 			 xmlns:local="using:AvaloniaApplication1"
>              x:Class="AvaloniaApplication1.UserControl1"
>              x:Name="root" >
> 	<UserControl.Styles>
> 		<Style Selector=":is(local|UserControl1)">
> 			<Setter Property="Template">
> 				<ControlTemplate>
> 					<Border x:Name="B">
> 						<ContentPresenter  Content="{TemplateBinding Content}"></ContentPresenter>
> 						<AdornerLayer.Adorner>
> 							<Grid IsVisible="{Binding #root.IsChecked}">
> 							</Grid>
> 						</AdornerLayer.Adorner>
> 					</Border>
> 				</ControlTemplate>
> 			</Setter>
> 		</Style>
> 	</UserControl.Styles>
> </UserControl>
> ```

Well, that's great! I am just trying a stupid way by using GetLogicalParent in code-behind to get UserControl. By your way is more simple and convenient Thank you.
